napari/layers/utils/array_interface.py

```diff
@@ -0,0 +1,21 @@
+import sys
+from typing import Tuple, Union
+
+import numpy as np
+
+if sys.version_info.minor < 8:
+    from typing_extensions import Protocol
+else:
+    from typing import Protocol
+
+
+class ArrayInterface(Protocol):
+    shape: Tuple[int] = (0,)
+    dtype: type = np.uint8
+    ndim: int = 1
+
+    def __getitem__(self, item) -> Union['ArrayInterface', int, float, bool]:
+        ...
+
+    def __len__(self) -> int:
+        ...
```

Comment:
@tlambert03 @jni Something like this?

Reply:
yep! something like that!  will be fun/interesting to try to "break" it with various cases... that's where i've generally run into hurdles.  lack of type stubs in various libraries also makes this a little less "practically useful" sometimes, which is another thing that has cause me to drag my feet on it.  But this is exactly how we should start.

Reply:
It might also be worth considering [subprotocols](https://mypy.readthedocs.io/en/stable/protocols.html#defining-subprotocols-and-subclassing-protocols).  For instance, there are lot of places where we have restrictions on the shape/type of an array (like points data), and we had imagined the ability to hint that.  Subprotocols will let us construct specific combinations,  For instance:

```python
class SupportsShape(Protocol):
    shape: Tuple[int, ...]

class Is2D(Protocol):
    shape: Tuple[int, int]

class SupportsLen(Protocol):
    def __len__(self, item) -> int: ...

class ArrayInterface(SupportsShape, SupportsLen, Protocol): ...
class Array2DInterface(Is2D, SupportsLen, Protocol): ...
PointsLayer = Union[Array2Dinterface, ...]  # or something like that
```

Reply:
I implement only minimal signature because I'm not familliar with other numpy like packets. 

So I think that best is to know where also (other than `set_image`) could be passed non numpy array and check which properties are needed.

Reply:
see now.